### ISSUE TITLE
Incorrect multiplication of Hamiltonians

### ISSUE THREAD

**Issue by trbromley** [PennyLaneAI/pennylane#1272]
#### Issue description

PennyLane can incorrectly multiply Hamiltonians:

```pycon
>>> h = qml.Hamiltonian([1, 1], [qml.PauliZ(0), qml.PauliZ(0) @ qml.PauliX(1)])
>>> h_sq = h @ h
>>> print(h_sq)
  (1) [Z0 Z0]
+ (3) [Z0 Z0 X1]
```
The expected result is `2 * I + 2 * X1`. Note that the coefficients are different, this is not an issue with insufficient simplification.
The problem appears to be due to terms in the Hamiltonian sharing the same wire. This replicates some of the discussion in these issues:
- https://github.com/PennyLaneAI/pennylane/issues/1107
- https://github.com/PennyLaneAI/pennylane/issues/715

However, even if it is a known issue that we can't have tensors composed of terms acting on the same wire, it still seems dangerous here to give an incorrect output from `h @ h` rather than, e.g., throwing an error.

#### Details

Looking at the `Hamiltonian.__matmul__()` code, everything works as expected until [this](https://github.com/PennyLaneAI/pennylane/blob/8fe31ad619222e2992130c5d036571f6ef7ab51c/pennylane/vqe/vqe.py#L296) line - the problem is with `Hamiltonian.simplify()`. In other words, we expect the following code to give `2 * I + 2 * X1` but it instead gives the result above.

```python
coeffs = [1, 1, 1, 1]
ops = [
    qml.PauliZ(0) @ qml.PauliZ(0),
    qml.PauliZ(0) @ qml.PauliZ(0) @ qml.PauliX(1),
    qml.PauliZ(0) @ qml.PauliX(1) @ qml.PauliZ(0),
    qml.PauliZ(0) @ qml.PauliX(1) @ qml.PauliZ(0) @ qml.PauliX(1)
]
h = qml.Hamiltonian(coeffs, ops, simplify=True)
```

The issue in `Hamiltonian.simplify()` is the use of `op.compare()` in [this](https://github.com/PennyLaneAI/pennylane/blob/8fe31ad619222e2992130c5d036571f6ef7ab51c/pennylane/vqe/vqe.py#L157) line:

```pycon
>>> op1 = qml.PauliZ(0) @ qml.PauliX(1) @ qml.PauliZ(0)
>>> op2 = qml.PauliZ(0) @ qml.PauliX(1) @ qml.PauliZ(0) @ qml.PauliX(1)
>>> op1.compare(op2)
True
```

This should not be the case. `op1` is effectively `X1` and `op2` is the identity. In turn, the issue with `Tensor.compare()` is the use of [Tensor._obs_data()](https://github.com/PennyLaneAI/pennylane/blob/8fe31ad619222e2992130c5d036571f6ef7ab51c/pennylane/operation.py#L992). Due to the use of `set()`, this method implicitly assumes that tensors are only ever composed of operators on different wires, which is not the case here.

TLDR: matrix multiplication of `qml.Hamiltonian` can result in instances where the Hamiltonian's operations are tensors composed of observables that act on the same wire. This is assumed to not be the case lower down in the `Tensor` class.

**Comment by trbromley** [comment]
Looking at the tests [here](https://github.com/PennyLaneAI/pennylane/blob/8fe31ad619222e2992130c5d036571f6ef7ab51c/tests/test_vqe.py#L380), I think it's maybe known that the two Hamiltonians can't share terms acting on the same wire :thinking: But I still think we should either raise an error or fix, because it's an easy thing to get caught out on.
